api/app/signals/apps/dataset/sources/shape.py

```diff
@@ -0,0 +1,111 @@
+import os
+import zipfile
+from urllib.parse import urlsplit
+
+import requests
+from django.contrib.gis.gdal import DataSource, OGRGeomType
+from django.contrib.gis.geos import MultiPolygon
+from django.db import transaction
+
+from signals.apps.dataset.base import AreaLoader
+from signals.apps.signals.models import Area, AreaType
+
+
+class ShapeBoundariesLoader(AreaLoader):
+    DATASET_INFO = {
+        'GENERIC': {}
+    }
+
+    PROVIDES = DATASET_INFO.keys()
+
+    def __init__(self, **options):
+        type_string = options['type_string']
+        directory = options['dir']
+        assert type_string in self.PROVIDES
+
+        self.area_type, _ = AreaType.objects.get_or_create(
+            name=options['type'],
+            code=options['type'],
+            description=f"{options['type']} data",
+        )
+        self.directory = directory  # Data downloaded / processed here. Caller is responsible to clean-up directory.
+        self.DATASET_URL = options['url']
+        self.data_file = options['shp']
+        self.code_field = options['code']
+        self.name_field = options['name']
+
+    def _download(self, zip_fullpath):
+        """
+        Download relevant data file.
+        """
+        if os.path.exists(zip_fullpath):
+            return  # Datafile already downloaded.
+
+        with requests.get(self.DATASET_URL, stream=True, verify=False) as r:
+            r.raise_for_status()
+            with open(zip_fullpath, 'wb') as f:
+                for chunk in r.iter_content(chunk_size=8192):
+                    f.write(chunk)
+
+    def _unzip(self, zip_fullpath):
+        """
+        Extract ZIP file to temp_dir.
+        """
+        with zipfile.ZipFile(zip_fullpath, 'r') as zf:
+            zf.extractall(path=self.directory)
+
+    def _load_shape_data(self, data_fullpath):
+        """
+        Load "gemeente", "wijk" or "buurt" areas from the CBS provided shapefiles.
```

Comment:
Rephrase this docstring, in this contex the code is more general than "gemeente", "wijk" or "buurt".

Reply:
check